### PR TITLE
Add more attributes for Cassandra to database semantic conventions

### LINKED ISSUE #1058: Additional Cassandra Attributes
**What are you trying to achieve?**
I'd like to define more cassandra attributes, as requested by @trask in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1314#issuecomment-703134133
Additionally, some of these tags may apply to other DBs as well.

What did you expect to see?

I expected to see more detailed information, including table name, query execution information, etc. Listed below.
```java
  public static final String QUERY_PAGE_SIZE = "query.pageSize";
  public static final String QUERY_CL = "query.cl"; // consistency level
  public static final String DB_HOST_DC = "db.host.dc";
  public static final String DB_HOST_ID = "db.host.id";
  public static final String QUERY_SPECULATIVE_EXECUTION_COUNT = "query.speculativeExecutionCount";
  public static final String QUERY_IDEMPOTENCE = "query.idempotence";
  public static final String QUERY_TABLE = "query.table";
```
**Additional context.**

Add any other context about the problem here. If you followed an existing documentation, please share the link to it.

Some of these tags already exist in the opentracing cassandra instrumentation https://github.com/opentracing-contrib/java-cassandra-driver/blob/master/opentracing-cassandra-driver-3/src/main/java/io/opentracing/contrib/cassandra/TracingSession.java#L500

### PR DESCRIPTION
**Pull request by FrankSpitulski**
Fixes #1058 

## Changes

Adds additional optional cassandra tags.

Related issues #

Related [instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1298)

### CODE REVIEW

File: semantic_conventions/trace/database.yaml [hunk 1]
```diff
@@ -251,6 +251,49 @@ groups:
...
+          tag: call-level-tech-specific
+          brief: >
+            The page size of the query.
+          examples: [5000]
+        - id: consistency_level
+          type: string
+          tag: call-level-tech-specific
+          brief: >
+            The consistency level of the query.
+          examples: 'LOCAL_QUORUM'
```
Comment (arminru) [inline semantic_conventions/trace/database.yaml:265]: This could be an enum (see the definition of `db.system` in line 8 above).

File: semantic_conventions/trace/database.yaml [hunk 2]
```diff
@@ -251,6 +251,49 @@ groups:
...
+        - id: table
+          type: string
+          tag: call-level-tech-specific
+          brief: >
+            The table being queried.
+          examples: 'mytable'
```
Comment (arminru) [inline semantic_conventions/trace/database.yaml:271]: We could consider using `db.sql.table` instead which has the same meaning and a more elaborate definition. CQL is close to SQL after all.

File: semantic_conventions/trace/database.yaml [hunk 3]
```diff
@@ -251,6 +251,49 @@ groups:
...
+          tag: call-level-tech-specific
+          brief: >
+            Whether or not the query is idempotent.
+          examples: [true, false]
+        - id: speculative_execution_count
+          type: number
+          required: always
+          tag: call-level-tech-specific
+          brief: >
+            The number of times a query was speculatively executed. A query that was not speculatively executed will have a value of 0.
```
Comment (arminru) [inline semantic_conventions/trace/database.yaml:283]:
```suggestion
        - id: speculative_execution_count
          type: number
          tag: call-level-tech-specific
          brief: >
            The number of times a query was speculatively executed. Not set or `0` if the query was not executed speculatively.
```

File: semantic_conventions/trace/database.yaml [hunk 4]
```diff
@@ -251,6 +251,49 @@ groups:
...
+          type: string
+          tag: call-level-tech-specific
+          brief: >
+            The host id reported by the coordinating node for a query.
+          examples: 'be13faa2-8574-4d71-926d-27f16cf8a7af'
+        - id: host.dc
+          type: string
+          tag: call-level-tech-specific
+          brief: >
+            The host dc reported by the coordinating node for a query.
```
Comment (arminru) [inline semantic_conventions/trace/database.yaml:295]:
```suggestion
            The host data center reported by the coordinating node for a query.
```

File: semantic_conventions/trace/database.yaml [hunk 5]
```diff
@@ -251,6 +251,49 @@ groups:
...
+          tag: call-level-tech-specific
+          brief: >
+            The table being queried.
+          examples: 'mytable'
+        - id: idempotence
+          type: boolean
+          tag: call-level-tech-specific
+          brief: >
+            Whether or not the query is idempotent.
+          examples: [true, false]
```
Comment (arminru) [inline semantic_conventions/trace/database.yaml:277]:
```suggestion
```
You could leave those out for booleans (it also looks a bit odd when rendered since the YAML parser normalizes it to its capital letter representation).

**Review by arminru — COMMENTED** [review]
Hi @FrankSpitulski!
While some SIGs use PR titles as the one here, this is not common in the spec repo. Apart from that, please use the correct term "attribute" or people might get confused. What about "Add more attributes for Cassandra to database semantic conventions"?
Also please add an entry in `CHANGELOG.md`.
Thanks!

File: semantic_conventions/trace/database.yaml (hunk 1 above)
Comment (Oberon00) [reply semantic_conventions/trace/database.yaml:265]: Indeed, this should be an enum. Or at least you should describe where the string comes from. This looks like a constant/enumerator defined in some library. But note that we want consistency across languages, so this is not necessarily the best source, e.g. one language might call it `LOCAL_QUORUM`, another one `kLocalQuorum` a third `CASSANDRA_CONSISTENCY_LOCAL_QUORUM`. So we should have definitive list of those here, or a link to a definitive list (e.g. Cassandra docs).

File: semantic_conventions/trace/database.yaml (hunk 2 above)
Comment (Oberon00) [reply semantic_conventions/trace/database.yaml:271]: I think having a separate attribute here for cassandra is fine. Otherwise we might want to rename `db.sql.table` to `db.table` after all and describe to which technologies it applies. But I think I prefer the separate attribute a bit.

File: semantic_conventions/trace/database.yaml (hunk 5 above)
Comment (Oberon00) [reply semantic_conventions/trace/database.yaml:277]: I think they are enforced by the parser. Or did we have an exception for bool? Maybe.

Comment (thisthat) [reply semantic_conventions/trace/database.yaml:277]: They are optional for boolean and enum: https://github.com/open-telemetry/opentelemetry-specification/blob/master/semantic_conventions/syntax.md#attributes

File: semantic_conventions/trace/database.yaml (hunk 2 above)
Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml:271]: +1 I agree that should either be separate or use a top level db attribute

File: semantic_conventions/trace/database.yaml (hunk 1 above)
Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml:265]: the docs have the full set https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/dml/dmlConfigConsistency.html

Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml]: changed to use an enum

File: semantic_conventions/trace/database.yaml [hunk 6]
```diff
@@ -251,6 +251,69 @@ groups:
...
+        - id: page_size
+          type: number
+          tag: call-level-tech-specific
+          brief: >
+            The page size of the query.
+          examples: [5000]
+        - id: consistency_level
+          tag: call-level-tech-specific
+          brief: >
+            The consistency level of the query. Based on consistency values from [cql](https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/dml/dmlConfigConsistency.html).
```
Comment (arminru) [inline semantic_conventions/trace/database.yaml:263]:
```suggestion
            The consistency level of the query. Based on consistency values from [CQL](https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/dml/dmlConfigConsistency.html).
```
Or did you mean to write `cqlsh`?

**Review by arminru — COMMENTED** [review]
LGTM!
It would be great if someone more familiar with Cassandra than me could take a look as well 🙂

File: specification/trace/semantic_conventions/database.md [hunk 7]
```diff
@@ -162,6 +162,13 @@ For example, when retrieving a document, `db.operation` would be set to (literal
 | Attribute  | Type | Description  | Example  | Required |
 |---|---|---|---|---|
 | `db.cassandra.keyspace` | string | The name of the keyspace being accessed. To be used instead of the generic `db.name` attribute. | `mykeyspace` | Yes |
+| `db.cassandra.page_size` | number | The page size of the query. | `5000` | No |
+| `db.cassandra.consistency_level` | string enum | The consistency level of the query. Based on consistency values from [cql](https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/dml/dmlConfigConsistency.html). | `ALL` | No |
+| `db.cassandra.table` | string | The table being queried. | `mytable` | No |
+| `db.cassandra.idempotence` | boolean | Whether or not the query is idempotent. |  | No |
+| `db.cassandra.speculative_execution_count` | number | The number of times a query was speculatively executed. Not set or `0` if the query was not executed speculatively. | `0`<br>`2` | No |
+| `db.cassandra.host.id` | string | The host id reported by the coordinating node for a query. | `be13faa2-8574-4d71-926d-27f16cf8a7af` | No |
```
Comment (yurishkuro) [inline specification/trace/semantic_conventions/database.md:170]: "host" is confusing here, how about "coordinator.id"?

File: specification/trace/semantic_conventions/database.md [hunk 8]
```diff
@@ -162,6 +162,13 @@ For example, when retrieving a document, `db.operation` would be set to (literal
 | Attribute  | Type | Description  | Example  | Required |
 |---|---|---|---|---|
 | `db.cassandra.keyspace` | string | The name of the keyspace being accessed. To be used instead of the generic `db.name` attribute. | `mykeyspace` | Yes |
+| `db.cassandra.page_size` | number | The page size of the query. | `5000` | No |
+| `db.cassandra.consistency_level` | string enum | The consistency level of the query. Based on consistency values from [cql](https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/dml/dmlConfigConsistency.html). | `ALL` | No |
+| `db.cassandra.table` | string | The table being queried. | `mytable` | No |
+| `db.cassandra.idempotence` | boolean | Whether or not the query is idempotent. |  | No |
```
Comment (yurishkuro) [inline specification/trace/semantic_conventions/database.md:168]: Who can determine whether a query is idempotent?  This does not sound like Cassandra specific.

File: semantic_conventions/trace/database.yaml [hunk 9]
```diff
@@ -251,6 +251,69 @@ groups:
           brief: >
             The name of the keyspace being accessed. To be used instead of the generic `db.name` attribute.
           examples: 'mykeyspace'
+        - id: page_size
```
Comment (yurishkuro) [inline semantic_conventions/trace/database.yaml:254]: the driver seems to call it "fetch size": https://docs.datastax.com/en/developer/java-driver/2.1/manual/paging/

The description below could be better, like "fetch size used for paging, i.e. how many rows will be returned at once"

File: specification/trace/semantic_conventions/database.md (hunk 8 above)
Comment (FrankSpitulski) [reply specification/trace/semantic_conventions/database.md:168]: Every query is marked as idempotent or not. https://docs.datastax.com/en/devapp/doc/devapp/driversQueryIdempotence.html

File: specification/trace/semantic_conventions/database.md (hunk 7 above)
Comment (FrankSpitulski) [reply specification/trace/semantic_conventions/database.md:170]: sure I can change it to coordinator

File: semantic_conventions/trace/database.yaml (hunk 6 above)
Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml:263]: CQL is fine. Either work. I'll use CQL.

File: semantic_conventions/trace/database.yaml (hunk 9 above)
Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml:254]: The most recent driver uses the term page size https://docs.datastax.com/en/developer/java-driver/4.9/manual/core/paging/

File: specification/performance-benchmark.md [hunk 10]
```diff
@@ -1,70 +1,70 @@
-# Performance Benchmark of OpenTelemetry API
```
Comment (yurishkuro) [inline specification/performance-benchmark.md:1]: can this file be reverted? doesn't seem related to the PR.

Comment (FrankSpitulski) [reply specification/performance-benchmark.md:1]: yeah this is weird. I don't remember touching this file.

File: semantic_conventions/trace/database.yaml [hunk 11]
```diff
@@ -251,6 +251,69 @@ groups:
...
+          type: number
+          tag: call-level-tech-specific
+          brief: >
+            The number of times a query was speculatively executed. Not set or `0` if the query was not executed speculatively.
+          examples: [0, 2]
+        - id: coordinator.id
+          type: string
+          tag: call-level-tech-specific
+          brief: >
+            The (node's) id of the coordinating node for a query.
```
Comment (arminru) [inline semantic_conventions/trace/database.yaml:309]:
```suggestion
            The ID of the coordinating node for a query.
```

File: specification/trace/semantic_conventions/database.md [hunk 12]
```diff
@@ -162,6 +162,13 @@ For example, when retrieving a document, `db.operation` would be set to (literal
 | Attribute  | Type | Description  | Example  | Required |
 |---|---|---|---|---|
 | `db.cassandra.keyspace` | string | The name of the keyspace being accessed. To be used instead of the generic `db.name` attribute. | `mykeyspace` | Yes |
+| `db.cassandra.page_size` | number | The fetch size used for paging, i.e. how many rows will be returned at once. | `5000` | No |
+| `db.cassandra.consistency_level` | string enum | The consistency level of the query. Based on consistency values from [CQL](https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/dml/dmlConfigConsistency.html). | `ALL` | No |
+| `db.cassandra.table` | string | The table being queried. | `mytable` | No |
+| `db.cassandra.idempotence` | boolean | Whether or not the query is idempotent. |  | No |
+| `db.cassandra.speculative_execution_count` | number | The number of times a query was speculatively executed. Not set or `0` if the query was not executed speculatively. | `0`<br>`2` | No |
+| `db.cassandra.coordinator.id` | string | The (node's) id of the coordinating node for a query. | `be13faa2-8574-4d71-926d-27f16cf8a7af` | No |
```
Comment (arminru) [inline specification/trace/semantic_conventions/database.md:170]:
```suggestion
| `db.cassandra.coordinator.id` | string | The ID of the coordinating node for a query. | `be13faa2-8574-4d71-926d-27f16cf8a7af` | No |
```

**Review by arminru — APPROVED** [review]
LGTM!
It would be great if someone more familiar with Cassandra than me could take a look as well 🙂

File: semantic_conventions/trace/database.yaml [hunk 13]
```diff
@@ -251,6 +251,69 @@ groups:
...
+            The fetch size used for paging, i.e. how many rows will be returned at once.
+          examples: [5000]
+        - id: consistency_level
+          tag: call-level-tech-specific
+          brief: >
+            The consistency level of the query. Based on consistency values from [CQL](https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/dml/dmlConfigConsistency.html).
+          type:
+            members:
+              - id: ALL
+                value: 'ALL'
```
Comment (Oberon00) [inline semantic_conventions/trace/database.yaml:267]: We usually use all-lowercase for these values. If there is any chance that we allow custom values in the future and these are the actual cassandra-native values (and not only used in the Java client lib), I'm fine with uppercase though.

File: semantic_conventions/trace/database.yaml (hunk 2 above)
Comment (SergeyKanzhelev) [reply semantic_conventions/trace/database.yaml:293]: `db.sql.table` is already mentioned in the rendered table. Having both is confusing:

File: semantic_conventions/trace/database.yaml (hunk 13 above)
Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml:267]: This would fall under being cassandra native values. There was some previous discussion on it here. https://github.com/open-telemetry/opentelemetry-specification/pull/1217#discussion_r520797155

**Comment by SergeyKanzhelev** [comment]
I am still concerned with this: https://github.com/open-telemetry/opentelemetry-specification/pull/1217#discussion_r524727177 two attributes with the similar semantic in the same table:

| Attribute  | Type | Description  | Example  | Required |
|---|---|---|---|---|
| `db.cassandra.table` | string | The table being queried. | `mytable` | No |
| `db.sql.table` | string | The name of the primary table that the operation is acting upon, including the schema name (if applicable). [2] | `public.users`<br>`customers` | Conditional<br>Recommended if available. |

File: semantic_conventions/trace/database.yaml (hunk 2 above)
Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml:293]: There isn't much more to say @SergeyKanzhelev unless you have a new suggestion.

Comment (SergeyKanzhelev) [reply semantic_conventions/trace/database.yaml:293]: the question is what having a cassandra-specific attribute should be defined? Why not use generic `db.sql.table`? If this is totally wrong, can we change the tool to at least not show both attributes in the same table?

Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml:293]: The issue with using db.sql.table is that it's not generic. It's defined in the sql specific section instead of in the generic DB section.

Comment (Oberon00) [reply semantic_conventions/trace/database.yaml:293]: Since we have quite a few of them, it would make sense to have a separate table for cassandra, indeed. This can be achieved by using a different tag, or a different semantic convention group (that extends the generic db)

Comment (Oberon00) [reply semantic_conventions/trace/database.yaml:293]: Also, the brief of `cassandra.table` should mention Cassandra and the brief of sql.table SQL.

Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml:293]: Bumping, will get to this eventually

Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml:293]: @Oberon00, I just had a closer look. It seems it already is a different convention group that extends the generic db.

```
groups:
    - id: db ...

    - id: db.mssql ...

    - id: db.cassandra
      prefix: db.cassandra
      extends: db
```

Comment (Oberon00) [reply semantic_conventions/trace/database.yaml:293]: Cool, then it should be possible to just edit the magic HTML comments in the markdown to generate that in a separate table.

Comment (FrankSpitulski) [reply semantic_conventions/trace/database.yaml]: @Oberon00 it's been split

File: specification/trace/semantic_conventions/database.md [hunk 14]
```diff
@@ -178,6 +179,25 @@ For example, when retrieving a document, `db.operation` would be set to (literal
 **[2]:** It is not recommended to attempt any client-side parsing of `db.statement` just to get this property, but it should be set if it is provided by the library being instrumented. If the operation is acting upon an anonymous table, or more than one table, this value MUST NOT be set.
 <!-- endsemconv -->
 
+#### Cassandra
+
+Seperated for clarity.
```
Comment (FrankSpitulski) [inline specification/trace/semantic_conventions/database.md:184]:
```suggestion
Separated for clarity.
```

**Comment by FrankSpitulski** [comment]
@SergeyKanzhelev It's been moved to a new table

**Review by SergeyKanzhelev — APPROVED** [review]
LGTM, but if somebody with the experience of running Cassandra can comment - would be great.

**Comment by iNikem** [comment]
@open-telemetry/technical-committee does anything prevent this from merging? Either we don't have "somebody with the experience of running Cassandra" or they do not object. Either way there is no point in waiting forever, is it?

File: specification/trace/semantic_conventions/database.md [hunk 15]
```diff
@@ -178,6 +179,25 @@ For example, when retrieving a document, `db.operation` would be set to (literal
 **[2]:** It is not recommended to attempt any client-side parsing of `db.statement` just to get this property, but it should be set if it is provided by the library being instrumented. If the operation is acting upon an anonymous table, or more than one table, this value MUST NOT be set.
 <!-- endsemconv -->
 
+#### Cassandra
+
+Separated for clarity.
+
+<!-- semconv db.tech(tag=call-level-tech-specific-cassandra) -->
```
Comment (Oberon00) [inline specification/trace/semantic_conventions/database.md:186]: I think this would be cleaner if you used

```suggestion
<!-- semconv db.cassandra -->
```

here. Then you could remove the `tag: call-level-tech-specific-cassandra` line everywhere.

Also, you can remove the `include:` for cassandra from `db.tech`.

File: specification/trace/semantic_conventions/database.md [hunk 16]
```diff
@@ -178,6 +179,25 @@ For example, when retrieving a document, `db.operation` would be set to (literal
 **[2]:** It is not recommended to attempt any client-side parsing of `db.statement` just to get this property, but it should be set if it is provided by the library being instrumented. If the operation is acting upon an anonymous table, or more than one table, this value MUST NOT be set.
 <!-- endsemconv -->
 
+#### Cassandra
+
+Separated for clarity.
+
```
Comment (Oberon00) [inline specification/trace/semantic_conventions/database.md:185]:
```suggestion
```

**Review by Oberon00 — APPROVED** [review]

**Comment by Oberon00** [comment]
IMHO: Let's merge it, and address my YAML nitpicks later. We can't make the review periods arbitrarily long and still expect contributors to respond.
